src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/ConstantExpectedTests.cs

```diff
@@ -0,0 +1,722 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Testing;
+using Microsoft.NetCore.Analyzers.Performance;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.CSharp.Analyzers.Performance.CSharpConstantExpectedAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+namespace Microsoft.CodeAnalysis.NetAnalyzers.UnitTests.Microsoft.NetCore.Analyzers.Performance
+{
+    public sealed class ConstantExpectedTests
+    {
+        [Theory]
+        [InlineData("char", "char.MinValue", "char.MaxValue")]
+        [InlineData("sbyte", "sbyte.MinValue", "sbyte.MaxValue")]
+        [InlineData("short", "short.MinValue", "short.MaxValue")]
+        [InlineData("int", "int.MinValue", "int.MaxValue")]
+        [InlineData("long", "long.MinValue", "long.MaxValue")]
+        [InlineData("nint", "int.MinValue", "int.MaxValue")]
+        [InlineData("byte", "byte.MinValue", "byte.MaxValue")]
+        [InlineData("ushort", "ushort.MinValue", "ushort.MaxValue")]
+        [InlineData("uint", "uint.MinValue", "uint.MaxValue")]
+        [InlineData("ulong", "ulong.MinValue", "ulong.MaxValue")]
+        [InlineData("nuint", "uint.MinValue", "uint.MaxValue")]
+        [InlineData("bool", "false", "true")]
+        [InlineData("float", "float.MinValue", "float.MaxValue")]
+        [InlineData("double", "double.MinValue", "double.MaxValue")]
+        public static async Task TestConstantExpectedSupportedUnmanagedTypesAsync(string type, string minValue, string maxValue)
+        {
+            string csInput = @$"
+using System;
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class Test
+{{
+    public static void TestMethod1([ConstantExpected] {type} val) {{ }}
+    public static void TestMethod2([ConstantExpected(Min={minValue})] {type} val) {{ }}
+    public static void TestMethod3([ConstantExpected(Max={maxValue})] {type} val) {{ }}
+    public static void TestMethod4([ConstantExpected(Min={minValue}, Max={maxValue})] {type} val) {{ }}
+    public static void TestMethod5([ConstantExpected(Min=null)] {type} val) {{ }}
+    public static void TestMethod6([ConstantExpected(Max=null)] {type} val) {{ }}
+    public static void TestMethod7([ConstantExpected(Min=null, Max=null)] {type} val) {{ }}
+}}
+";
+            await TestCSAsync(csInput);
+        }
+
+        [Fact]
+        public static async Task TestConstantExpectedSupportedComplexTypesAsync()
+        {
+            string csInput = @"
+using System;
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class Test
+{
+    public static void TestMethodString([ConstantExpected] string val) { }
```

Comment:
if I remember correctly only value types are allowed, excluding string type

Correct me if I am wrong @tannergooding

Reply:
iirc all primitives except object, dynamic, nint, nuint.

-> Half is also a value type

Reply:
Only types that can have constant values (so the primitive types, decimal, and string)